webapp/src/js/services/update-service-worker.js

```diff
@@ -0,0 +1,50 @@
+/*
+Handles service worker updates
+*/
+angular.module('inboxServices').factory('UpdateServiceWorker', function($window, $log, $timeout) {
+  'use strict';
+
+  const retryFailedUpdateAfterSec = 5 * 60;
+  let existingUpdateLoop;
+
+  function update(onSuccess) {
+    // This avoids multiple updates retrying in parallel
+    if (existingUpdateLoop) {
+      $timeout.cancel(existingUpdateLoop);
+      existingUpdateLoop = undefined;
+    }
+
+    $window.navigator.serviceWorker.getRegistrations()
+      .then(function(registrations) {
+        const registration = registrations && registrations.length && registrations[0];
+        if (!registration) {
+          $log.warn('Cannot update service worker - no active workers found');
+          return;
+        }
+
+        registration.onupdatefound = function() {
+          const installingWorker = registration.installing;
+          installingWorker.onstatechange = function() {
+            switch (installingWorker.state) {
+              case 'activated':
+                $log.info('New service worker activated');
+                registration.onupdatefound = undefined;
+                onSuccess();
+                break;
+              case 'redundant':
+                $log.warn(`Service worker failed to install or marked as redundant. Retrying install in ${retryFailedUpdateAfterSec}secs.`);
+                existingUpdateLoop = $timeout(() => update(onSuccess), retryFailedUpdateAfterSec * 1000);
+                registration.onupdatefound = undefined;
+                break;
+              default:
+                $log.debug(`Service worker state changed to ${installingWorker.state}!`);
+            }
+          };
+        };
+
+        registration.update();
+      });
+  }
+
+  return update;
```

Comment:
This looks very similar to the code in the bootstrap for registering the service worker - can the common bits be extracted into shared code somewhere, maybe `/webapp/src/js/modules/...`?

Reply:
I agree they are similar. I tried having them share code previously and I remember thinking it wasn't any cleaner. 

Note update.js is calling `navigator.serviceWorker.getRegistrations` and swRegister.js is calling `navigator.serviceWorker.register`. 

The `onupdatefound` block is basically all that can be abstracted, but after wiring up the multiple callbacks -- it is maybe more trouble than it is worth. Happy to take another try if you still feel there is sufficient shared code here.

Reply:
Ok, that's fine. Thanks for looking into it!